**Project Description**

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

We've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

# **Description of the data:**
<br>
<mark>books</mark><br>
Contains data on books:<br>
- book_id<br>
- author_id<br>
- title<br>
- num_pages - number of pages<br>
- publication_date<br>
- publisher_id<br><br>
<mark>authors</mark><br>
Contains data on authors:<br>
- author_id
- author<br><br>
<mark>publishers</mark><br>
Contains data on publishers:<br>
- publisher_id<br>
- publisher<br><br>
<mark>ratings</mark><br>
Contains data on user_ratings:<br>
- rating_id<br>
- book_id<br>
- username - the name of the user who rated the book<br>
- rating <br><br>
<mark>reviews</mark><br>
Contains data on customer_reviews:<br>
- review_id<br>
- book_id<br>
- username - the name of the user who reviewed the book<br>
- text - the text of the revies

# **Task**

- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.

# **Instructions for completing the task**<a id='instructions'></a>
[Study the table (print the first rows)](#study)<br>
[Make an SQL query for each of the tasks](#query)<br>
[Describe your conclusions for each of the tasks](#describe)<br>

# **Access to the database**

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
pip install psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [3]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Study the table (print the first rows):<a id='study'></a>

**<mark>checking what columns we have in the data:</mark>**

In [4]:
engine.table_names()

['orders',
 'visits',
 'advertisment_costs',
 'publishers',
 'authors',
 'reviews',
 'ratings',
 'books']

**<mark>orders:</mark>**

In [5]:
orders = f'''SELECT *
FROM orders
LIMIT 5;'''
pd.io.sql.read_sql(orders, con = engine)

id               buyts revenue                   uid
0   1 2017-06-01 00:10:00  $17.00  10329302124590727494
1   2 2017-06-01 00:25:00   $0.55  11627257723692907447
2   3 2017-06-01 00:27:00   $0.37  17903680561304213844
3   4 2017-06-01 00:29:00   $0.55  16109239769442553005
4   5 2017-06-01 07:58:00   $0.37  14200605875248379450

In [6]:
orders_info = f'''SELECT *
FROM orders;'''
orders_info = pd.io.sql.read_sql(orders_info, con = engine)
orders_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50415 entries, 0 to 50414
Data columns (total 4 columns):
id         50415 non-null int64
buyts      50415 non-null datetime64[ns]
revenue    50415 non-null object
uid        50415 non-null object
dtypes: datetime64[ns](1), int64(1), object(2)
memory usage: 1.5+ MB


In [7]:
orders_describe = f'''SELECT *
FROM orders;'''
orders_describe = pd.io.sql.read_sql(orders_describe, con = engine)
orders_describe.describe(include = 'all')

id                buyts revenue                  uid
count   50415.000000                50415   50415                50415
unique           NaN                45991    1149                36523
top              NaN  2018-05-31 10:13:00   $1.83  3644482766749211722
freq             NaN                    9    2552                  239
first            NaN  2017-06-01 00:10:00     NaN                  NaN
last             NaN  2018-06-01 00:02:00     NaN                  NaN
mean    25208.000000                  NaN     NaN                  NaN
std     14553.701247                  NaN     NaN                  NaN
min         1.000000                  NaN     NaN                  NaN
25%     12604.500000                  NaN     NaN                  NaN
50%     25208.000000                  NaN     NaN                  NaN
75%     37811.500000                  NaN     NaN                  NaN
max     50415.000000                  NaN     NaN                  NaN

**Conclusion:** The table 'orders' has 4 columns and 50415 rows and a lot of missing values.

**<mark>visits:</mark>**

In [8]:
visits = f'''SELECT *
FROM visits
LIMIT 5'''
pd.io.sql.read_sql(visits, con = engine)

id                   uid   device               endts  sourceid  \
0   0  16879256277535980062    touch 2017-12-20 17:38:00         4   
1   1    104060357244891740  desktop 2018-02-19 17:21:00         2   
2   2   7459035603376831527    touch 2017-07-01 01:54:00         5   
3   3  16174680259334210214  desktop 2018-05-20 11:23:00         9   
4   4   9969694820036681168  desktop 2017-12-27 14:06:00         3   

              startts  
0 2017-12-20 17:20:00  
1 2018-02-19 16:53:00  
2 2017-07-01 01:54:00  
3 2018-05-20 10:59:00  
4 2017-12-27 14:06:00

In [9]:
visits_info = f'''SELECT *
FROM visits;'''
visits_info = pd.io.sql.read_sql(visits_info, con = engine)
visits_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 358532 entries, 0 to 358531
Data columns (total 6 columns):
id          358532 non-null int64
uid         358532 non-null object
device      358532 non-null object
endts       358532 non-null datetime64[ns]
sourceid    358532 non-null int64
startts     358532 non-null datetime64[ns]
dtypes: datetime64[ns](2), int64(2), object(2)
memory usage: 16.4+ MB


In [10]:
visits_describe = f'''SELECT *
FROM visits;'''
visits_describe = pd.io.sql.read_sql(visits_describe, con = engine)
visits_describe.describe(include = 'all')

id                   uid   device                endts  \
count   358532.000000                358532   358532               358532   
unique            NaN                228169        2               223894   
top               NaN  12869801667763685675  desktop  2017-11-24 16:51:00   
freq              NaN                   893   261913                   23   
first             NaN                   NaN      NaN  2017-06-01 00:02:00   
last              NaN                   NaN      NaN  2018-06-01 01:26:00   
mean    179265.500000                   NaN      NaN                  NaN   
std     103499.417694                   NaN      NaN                  NaN   
min          0.000000                   NaN      NaN                  NaN   
25%      89632.750000                   NaN      NaN                  NaN   
50%     179265.500000                   NaN      NaN                  NaN   
75%     268898.250000                   NaN      NaN                  NaN   
max     358531.000000                   NaN      NaN                  NaN   

             sourceid              startts  
count   358532.000000               358532  
unique            NaN               223977  
top               NaN  2017-11-24 16:06:00  
freq              NaN                   19  
first             NaN  2017-06-01 00:01:00  
last              NaN  2018-05-31 23:59:00  
mean         3.751551                  NaN  
std          1.916853                  NaN  
min          1.000000                  NaN  
25%          3.000000                  NaN  
50%          4.000000                  NaN  
75%          5.000000                  NaN  
max         10.000000                  NaN

**Conclusion:** The table 'visits' has 6 columns, 358532 rows. 

**<mark>advertisment_costs:</mark>**

In [11]:
advertisment_costs = f'''SELECT *
FROM advertisment_costs
LIMIT 5'''
pd.io.sql.read_sql(advertisment_costs, con = engine)

id  sourceid         dt   costs
0   1         1 2017-06-01  $75.20
1   2         1 2017-06-02  $62.25
2   3         1 2017-06-03  $36.53
3   4         1 2017-06-04  $55.00
4   5         1 2017-06-05  $57.08

In [12]:
advertisements_costs_info = f'''SELECT *
FROM advertisment_costs;'''
advertisements_costs_info = pd.io.sql.read_sql(advertisements_costs_info, con = engine)
advertisements_costs_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2542 entries, 0 to 2541
Data columns (total 4 columns):
id          2542 non-null int64
sourceid    2542 non-null int64
dt          2542 non-null datetime64[ns]
costs       2542 non-null object
dtypes: datetime64[ns](1), int64(2), object(1)
memory usage: 79.6+ KB


In [13]:
advertisements_costs_describe = f'''SELECT *
FROM advertisment_costs;'''
advertisements_costs_describe = pd.io.sql.read_sql(advertisements_costs_describe, con = engine)
advertisements_costs_describe.describe(include = 'all')

id     sourceid                   dt   costs
count   2542.000000  2542.000000                 2542    2542
unique          NaN          NaN                  364    2396
top             NaN          NaN  2018-03-03 00:00:00  $10.32
freq            NaN          NaN                    7       3
first           NaN          NaN  2017-06-01 00:00:00     NaN
last            NaN          NaN  2018-05-31 00:00:00     NaN
mean    1271.500000     4.857199                  NaN     NaN
std      733.956516     3.181581                  NaN     NaN
min        1.000000     1.000000                  NaN     NaN
25%      636.250000     2.000000                  NaN     NaN
50%     1271.500000     4.000000                  NaN     NaN
75%     1906.750000     9.000000                  NaN     NaN
max     2542.000000    10.000000                  NaN     NaN

**Conclusion:** The table 'advertisements_costs' has 4 columns and 2542 rows.

**<mark>publishers:</mark>**

In [14]:
publishers = f'''SELECT *
FROM publishers
LIMIT 5'''
pd.io.sql.read_sql(publishers, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [15]:
publishers_info = f'''SELECT *
FROM publishers;'''
publishers_info = pd.io.sql.read_sql(publishers_info, con = engine)
publishers_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [16]:
publishers_describe = f'''SELECT *
FROM publishers;'''
publishers_describe = pd.io.sql.read_sql(publishers_describe, con = engine)
publishers_describe.describe(include = 'all')

publisher_id                 publisher
count     340.000000                       340
unique           NaN                       340
top              NaN  Tyndale House Publishers
freq             NaN                         1
mean      170.500000                       NaN
std        98.293777                       NaN
min         1.000000                       NaN
25%        85.750000                       NaN
50%       170.500000                       NaN
75%       255.250000                       NaN
max       340.000000                       NaN

**Conclusion:** The table 'publishers' has 2 columns and 340 rows. 

**<mark>authors:</mark>**

In [17]:
authors = f'''SELECT *
FROM authors
LIMIT 5'''
pd.io.sql.read_sql(authors, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [18]:
authors_info = f'''SELECT *
FROM authors;'''
authors_info = pd.io.sql.read_sql(authors_info, con = engine)
authors_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [19]:
authors_describe = f'''SELECT *
FROM authors;'''
authors_describe = pd.io.sql.read_sql(authors_describe, con = engine)
authors_describe.describe(include = 'all')

author_id                           author
count   636.000000                              636
unique         NaN                              636
top            NaN  Theo LeSieg/Dr. Seuss/Roy McKie
freq           NaN                                1
mean    318.500000                              NaN
std     183.741666                              NaN
min       1.000000                              NaN
25%     159.750000                              NaN
50%     318.500000                              NaN
75%     477.250000                              NaN
max     636.000000                              NaN

**Conclusion:** The table on authors has 636 rows and 2 columns.

**<mark>reviews:</mark>**

In [20]:
reviews = f'''SELECT *
FROM reviews
LIMIT 5'''
pd.io.sql.read_sql(reviews, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [21]:
reviews_info = f'''SELECT *
FROM reviews;'''
reviews_info = pd.io.sql.read_sql(reviews_info, con = engine)
reviews_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [22]:
reviews_describe = f'''SELECT *
FROM reviews;'''
reviews_describe = pd.io.sql.read_sql(reviews_describe, con = engine)
reviews_describe.describe(include = 'all')

review_id      book_id username  \
count   2793.000000  2793.000000     2793   
unique          NaN          NaN      160   
top             NaN          NaN  susan85   
freq            NaN          NaN       29   
mean    1397.000000   504.693161      NaN   
std      806.413976   288.472931      NaN   
min        1.000000     1.000000      NaN   
25%      699.000000   259.000000      NaN   
50%     1397.000000   505.000000      NaN   
75%     2095.000000   753.000000      NaN   
max     2793.000000  1000.000000      NaN   

                                                     text  
count                                                2793  
unique                                               2793  
top     Lose conference under so anything. Must comput...  
freq                                                    1  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

**Conclusion:** The table on reviews has 4 columns and 2793 rows.

**<mark>ratings:</mark>**

In [23]:
ratings = f'''SELECT *
FROM ratings
LIMIT 5'''
pd.io.sql.read_sql(ratings, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [24]:
ratings_info = f'''SELECT *
FROM ratings;'''
ratings_info = pd.io.sql.read_sql(ratings_info, con = engine)
ratings_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [25]:
ratings_describe = f'''SELECT *
FROM ratings;'''
ratings_describe = pd.io.sql.read_sql(ratings_describe, con = engine)
ratings_describe.describe(include = 'all')

rating_id      book_id    username       rating
count   6456.000000  6456.000000        6456  6456.000000
unique          NaN          NaN         160          NaN
top             NaN          NaN  martinadam          NaN
freq            NaN          NaN          56          NaN
mean    3228.500000   510.574195         NaN     3.928284
std     1863.831001   284.141636         NaN     0.943303
min        1.000000     1.000000         NaN     1.000000
25%     1614.750000   291.000000         NaN     3.000000
50%     3228.500000   506.000000         NaN     4.000000
75%     4842.250000   750.000000         NaN     5.000000
max     6456.000000  1000.000000         NaN     5.000000

**Conclusion:** The table of ratings has 4 columns and 6456 rows.

**<mark>books:</mark>**

In [26]:
books = f'''SELECT *
FROM books
LIMIT 5'''
pd.io.sql.read_sql(books, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [27]:
books_info = f'''SELECT *
FROM books;'''
books_info = pd.io.sql.read_sql(books_info, con = engine)
books_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [28]:
books_describe = f'''SELECT *
FROM books;'''
books_describe = pd.io.sql.read_sql(books_describe, con = engine)
books_describe.describe(include = 'all')

book_id    author_id                title   num_pages  \
count   1000.000000  1000.000000                 1000  1000.00000   
unique          NaN          NaN                  999         NaN   
top             NaN          NaN  Memoirs of a Geisha         NaN   
freq            NaN          NaN                    2         NaN   
mean     500.500000   320.417000                  NaN   389.11100   
std      288.819436   181.620172                  NaN   229.39014   
min        1.000000     1.000000                  NaN    14.00000   
25%      250.750000   162.750000                  NaN   249.00000   
50%      500.500000   316.500000                  NaN   352.00000   
75%      750.250000   481.000000                  NaN   453.00000   
max     1000.000000   636.000000                  NaN  2690.00000   

       publication_date  publisher_id  
count              1000   1000.000000  
unique              618           NaN  
top          2004-06-01           NaN  
freq                 10           NaN  
mean                NaN    171.270000  
std                 NaN     99.082685  
min                 NaN      1.000000  
25%                 NaN     83.000000  
50%                 NaN    177.500000  
75%                 NaN    258.000000  
max                 NaN    340.000000

**Conclusion:** The table on books has 6 columns and 1000 rows.

# Make an SQL query for each of the tasks: <a id='query'></a>

- [Find the number of books released after January 1, 2000.](#january)
- [Find the number of user reviews and the average rating for each book.](#eachbook)
- [Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).](#publisher)
- [Identify the author with the highest average book rating (look only at books with at least 50 ratings).](#author)
- [Find the average number of text reviews among users who rated more than 50 books.](#reviews)

**FIND THE NUMBER OF BOOKS RELEASED AFTER JANUARY 1, 2000:**<a id='january'></a>

In [29]:
n_books_after_2000 = f'''SELECT COUNT(DISTINCT book_id) AS cnt_distinct_book_id
FROM books
WHERE publication_date > '2000-01-01';'''
print('Ther number of books released after January 1, 2000:')
pd.io.sql.read_sql(n_books_after_2000, con = engine)

Ther number of books released after January 1, 2000:


cnt_distinct_book_id
0                   819

**Conclusion:** As the task states, we chose '>' instead of '>=' and the number is 819.

**FIND THE NUMBER OF USER REVIEWS AND THE AVERAGE RATING FOR EACH BOOK:**<a id = 'eachbook'></a>

In [30]:
together = f'''SELECT 
books.book_id, books.title, COUNT(DISTINCT reviews.review_id) AS num_reviews, AVG(ratings.rating) AS avg_rating
FROM books
LEFT JOIN ratings
  ON books.book_id = ratings.book_id
LEFT JOIN reviews
ON books.book_id = reviews.book_id
GROUP BY books.book_id
ORDER BY COUNT(DISTINCT reviews.review_id) DESC
LIMIT 20;'''
print('The number of user reviews AND the average rating for each book ordered by number of reviews:')
pd.io.sql.read_sql(together, con = engine)

The number of user reviews AND the average rating for each book ordered by number of reviews:


book_id                                              title  num_reviews  \
0       948                            Twilight (Twilight  #1)            7   
1       627                                      The Alchemist            6   
2       497                          Outlander (Outlander  #1)            6   
3       750                The Hobbit  or There and Back Again            6   
4       207                                    Eat  Pray  Love            6   
5       656                                     The Book Thief            6   
6       695  The Curious Incident of the Dog in the Night-Time            6   
7       673                             The Catcher in the Rye            6   
8       733                          The Giver (The Giver  #1)            6   
9       734                                   The Glass Castle            6   
10      779  The Lightning Thief (Percy Jackson and the Oly...            6   
11      854                                           The Road            6   
12      963                                Water for Elephants            6   
13      696             The Da Vinci Code (Robert Langdon  #2)            6   
14      299  Harry Potter and the Chamber of Secrets (Harry...            6   
15      302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
16       15     A Clash of Kings  (A Song of Ice and Fire  #2)            5   
17      232                Extremely Loud and Incredibly Close            5   
18      267  Freakonomics: A Rogue Economist Explores the H...            5   
19      194                                            Dracula            5   

    avg_rating  
0     3.662500  
1     3.789474  
2     4.125000  
3     4.125000  
4     3.395833  
5     4.264151  
6     4.081081  
7     3.825581  
8     3.750000  
9     4.206897  
10    4.080645  
11    3.772727  
12    3.977273  
13    3.830508  
14    4.287500  
15    4.414634  
16    4.391304  
17    3.416667  
18    3.909091  
19    3.296296

In [31]:
together_ = f'''SELECT 
books.book_id, books.title, COUNT(DISTINCT reviews.review_id) AS num_reviews, AVG(ratings.rating) AS avg_rating
FROM books
LEFT JOIN ratings
  ON books.book_id = ratings.book_id
LEFT JOIN reviews
ON books.book_id = reviews.book_id
GROUP BY books.book_id
ORDER BY AVG(ratings.rating) DESC
LIMIT 20;'''
print('The number of user reviews AND the average rating for each book ordered by average rating:')
pd.io.sql.read_sql(together_, con = engine)

The number of user reviews AND the average rating for each book ordered by average rating:


book_id                                              title  num_reviews  \
0       347  In the Hand of the Goddess (Song of the Liones...            2   
1       390                                    Light in August            2   
2       275                                          Geek Love            2   
3       321                                Homage to Catalonia            2   
4       224                                      Evening Class            2   
5       330  How to Be a Domestic Goddess: Baking and the A...            1   
6       136  Captivating: Unveiling the Mystery of a Woman'...            2   
7       229           Evil Under the Sun (Hercule Poirot  #24)            2   
8       297                                         Hard Times            2   
9        62                                      Alas  Babylon            2   
10      182                                         Dead Souls            2   
11      169  Crucial Conversations: Tools for Talking When ...            2   
12      213                      Emily of New Moon (Emily  #1)            2   
13       76                                        Angels Fall            2   
14       57                    Act of Treason (Mitch Rapp  #9)            2   
15       17                      A Dirty Job (Grim Reaper  #1)            4   
16       55          A Woman of Substance (Emma Harte Saga #1)            2   
17       86      Arrows of the Queen (Heralds of Valdemar  #1)            2   
18       20              A Fistful of Charms (The Hollows  #4)            2   
19      418                                              March            2   

    avg_rating  
0          5.0  
1          5.0  
2          5.0  
3          5.0  
4          5.0  
5          5.0  
6          5.0  
7          5.0  
8          5.0  
9          5.0  
10         5.0  
11         5.0  
12         5.0  
13         5.0  
14         5.0  
15         5.0  
16         5.0  
17         5.0  
18         5.0  
19         5.0

**Conclusion:** We can see that the number of reviews on books does not directly correlate with the average rating on these books. 'Twilight' has the highest number of reviews.

**IDENTIFY THE PUBLISHER THAT HAS RELEASED THE GREATEST NUMBER OF BOOKS WITH MORE THAN 50 PAGES:**<BR>
**this will help us exclude brochures and similar publications from your analysis**<a id='publisher'></a>

In [32]:
publisher_greatest = f'''SELECT publisher, COUNT(DISTINCT books.book_id) AS num_books, MIN(DISTINCT books.num_pages) AS min_pages
FROM publishers
LEFT JOIN books
  ON publishers.publisher_id = books.publisher_id
WHERE books.num_pages > 50
GROUP BY publisher
ORDER BY COUNT(books.book_id) DESC
LIMIT 1;'''
print('Publisher with the highest number of published books with more than 50 pages:')
pd.io.sql.read_sql(publisher_greatest, con = engine)

Publisher with the highest number of published books with more than 50 pages:


publisher  num_books  min_pages
0  Penguin Books         42         80

**Conclusion:** The requested publisher is Penguin Books.

**IDENTIFY THE AUTHOR WITH THE HIGHEST AVERAGE BOOK RATING:**<br>
**look only at books with at least 50 ratings**<a id='author'></a>

In [33]:
author = f'''SELECT
  books.book_id,
  AVG(ratings.rating) AS avg_rating,
  COUNT(DISTINCT ratings.rating_id) AS number_ratings,
  authors.author AS author
FROM books
LEFT JOIN ratings
  ON books.book_id = ratings.book_id
LEFT JOIN authors
  ON books.author_id = authors.author_id
GROUP BY books.book_id, author
HAVING COUNT(DISTINCT ratings.rating_id) >= 50
ORDER BY AVG(ratings.rating) DESC
LIMIT 1;'''
print('The author with the highest average book rating (for books with more than 50 ratings):')
pd.io.sql.read_sql(author, con = engine)

The author with the highest average book rating (for books with more than 50 ratings):


book_id  avg_rating  number_ratings                      author
0      302    4.414634              82  J.K. Rowling/Mary GrandPré

**Conclusion:** The needed authors are J.K. Rowling/Mary GrandPré duo.

**FIND THE AVERAGE NUMBER OF TEXT REVIEWS AMONG USERS WHO RATED MORE THAN 50 BOOKS:**<a id='reviews'></a>

In [34]:
bookid = f'''SELECT
AVG(SUBQUERY.n_reviews) AS avg_text_rev
FROM
    (SELECT 
    username,
    COUNT(DISTINCT review_id) AS n_reviews
    FROM reviews
    WHERE username IN
        (SELECT username AS username
        FROM ratings
        GROUP BY username
        HAVING COUNT(DISTINCT book_id) > 50)
        GROUP BY username) AS SUBQUERY;'''

print('The average number of text reviews among users who rated more than 50 books:')
pd.io.sql.read_sql(bookid, con = engine)

The average number of text reviews among users who rated more than 50 books:


avg_text_rev
0     24.333333

**Conclusion**: The average number of text reviews among users who rated more than 50 books is 24.

# Describe your conclusions for each of the tasks: <a id='describe'></a>

- Find the number of books released after January 1, 2000: **819**
- Identify the publisher that has released the greatest number of books with more than 50 pages: **Penguin Books**
- Identify the author with the highest average book rating (look only at books with at least 50 ratings): **J.K. Rowling/Mary GrandPré**
- Find the average number of text reviews among users who rated more than 50 books: **24**

Thank you for checking!